langchain/src/chains/combine_docs_chain.ts

```diff
@@ -215,3 +218,152 @@ export class MapReduceDocumentsChain
     };
   }
 }
+
+export interface RefineDocumentsChainInput extends StuffDocumentsChainInput {
+  refineLLMChain: LLMChain;
+  documentPrompt: BasePromptTemplate;
+}
+
+/**
+ * Combine documents by doing a first pass and then refining on more documents.
+ * @augments BaseChain
+ * @augments RefineDocumentsChainInput
+ */
+export class RefineDocumentsChain
+  extends BaseChain
+  implements RefineDocumentsChainInput
+{
+  llmChain: LLMChain;
+
+  inputKey = "input_documents";
+
+  outputKey = "output_text";
+
+  documentVariableName = "context";
+
+  initialResponseName = "existing_answer";
+
+  refineLLMChain: LLMChain;
+
+  get defaultDocumentPrompt(): BasePromptTemplate {
+    return new PromptTemplate({
+      inputVariables: ["page_content"],
+      template: "{page_content}",
+    });
+  }
+
+  documentPrompt = this.defaultDocumentPrompt;
+
+  get inputKeys() {
+    return [this.inputKey, ...this.refineLLMChain.inputKeys];
+  }
+
+  constructor(fields: {
+    llmChain: LLMChain;
+    refineLLMChain: LLMChain;
+    inputKey?: string;
+    outputKey?: string;
+    documentVariableName?: string;
+    documentPrompt?: BasePromptTemplate;
+    initialResponseName?: string;
+  }) {
+    super();
+    this.llmChain = fields.llmChain;
+    this.refineLLMChain = fields.refineLLMChain;
+    this.documentVariableName =
+      fields.documentVariableName ?? this.documentVariableName;
+    this.inputKey = fields.inputKey ?? this.inputKey;
+    this.documentPrompt = fields.documentPrompt ?? this.documentPrompt;
+    this.initialResponseName =
+      fields.initialResponseName ?? this.initialResponseName;
+  }
+
+  _constructInitialInputs(docs: Document[], rest: Record<string, unknown>) {
```

Comment:
uber nit: maybe just pass only `docs[0]` into this function since thats all thats needed